packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -242,6 +218,18 @@ exports.sourceNodes = async (
     )
   }
 
+  // Add referenced values
+  for (let [key, references] of foreignReferenceMap) {
+    const node = getNode(key)
+    if (node) {
+      for (const field in references) {
+        node[field] = (node[field] || []).concat(
+          references[field].filter(reference => !!getNode(reference))
+        )
```

Comment:
I think this reads better but perhaps I'm wrong. (Also lazily creates the array when necessary, rather than always, in case that's important). Of course, if the `refs` size is real-world potentially in the order of thousands or more then spread-push is probably not the best performant way.

```suggestion
        const refs = references[field].filter(reference => !!getNode(reference))
        if (refs.length) {
          if (!node[field]) node[field] = []
          node[field].push(...refs)
        )
```

Reply:
(I think you can get away with `references[field].filter(getNode)` but I do prefer the above since it's more obvious what's going on and doesn't risk passing on an accidentally relevant second parameter, index.)

Reply:
So maybe

```suggestion
        if (!node[field]) {
          node[field] = []
        }
        references[field].forEach(reference => {
          if (getNode(reference)) {
            node[field].push(reference); 
          }
        })
```